GVRf/Extensions/3DCursor/3DCursorLibrary/src/main/java/org/gearvrf/io/cursor3d/MovableObject.java

```diff
@@ -0,0 +1,213 @@
+import org.joml.Quaternionf;
+import org.joml.Vector3f;
+
+import java.io.IOException;
+import java.util.Map;
+
+/**
+ * This class defines a {@link SelectableObject} that is movable by any {@link Cursor}.
+ */
+public class MovableObject extends SelectableObject {
+    public static final String TAG = MovableObject.class.getSimpleName();
+
+    private Vector3f prevCursorPosition;
+    private Quaternionf rotation;
+    private Vector3f cross;
+
+    private GVRSceneObject selected;
+    private GVRSceneObject cursorSceneObject;
+    private Cursor cursor;
+    private CursorManager cursorManager;
+
+    /**
+     * Creates a {@link MovableObject} with the given {@link GVRSceneObject}. Such a
+     * {@link MovableObject} does not have a different {@link GVRSceneObject} for each
+     * {@link ObjectState}.
+     *
+     * @param sceneObject   The
+     * {@link GVRSceneObject} to be associated with the {@link MovableObject}
+     * @param cursorManager The {@link CursorManager} instance
+     */
+    public MovableObject(GVRSceneObject sceneObject, CursorManager cursorManager) {
+        super(sceneObject);
+        initialize(cursorManager);
+    }
+
+    /**
+     * Creates a {@link MovableObject} with a model that has a specific hierarchy
+     *                       root
+     *                    /  /  \  \
+     *                  /   /    \   \
+     *                /    /      \    \
+     *              /     /        \     \
+     *        default   behind colliding clicked
+     * This model has a root node at the top and a child for each of the states of the
+     * {@link MovableObject}. It is recommended that the different nodes representing  different
+     * {@link ObjectState} share a common set of vertices if possible.
+     *
+     * @param gvrContext    the {@link GVRContext}
+     * @param modelFileName A filename, relative to the {@code assets} directory. The file
+     *                      can be in a sub-directory of the {@code assets} directory:
+     *                      {@code "foo/bar.png"} will open the file
+     *                      {@code assets/foo/bar.png}
+     * @param objectStates  An array of the {@link ObjectState}s in the same order as the
+     *                      children of the
+     *                      root node of the model. This gives a mapping between the nodes in the
+     *                      model
+     *                      with the {@link ObjectState}s that they want to represent. This array
+     *                      should
+     *                      contain the {@link ObjectState#DEFAULT} element, not having it would
+     *                      result
+     *                      in an IllegalArgumentException
+     * @param cursorManager The {@link CursorManager} instance
+     * @throws IOException If the model file cannot be loaded.
+     */
+    public MovableObject(GVRContext gvrContext, String modelFileName, ObjectState[] objectStates,
```

Comment:
You could simplify using a default constructor that assumes objectStates as ObjectState[]{ObjectState.DEFAULT, ObjectState.BEHIND, ObjectState.COLLIDING, ObjectState.CLICKED}